numpy/core/numeric.py

```diff
@@ -662,14 +662,21 @@ def flatnonzero(a):
     return np.nonzero(np.ravel(a))[0]
 
 
-_mode_from_name_dict = {'v': 0,
-                        's': 1,
-                        'f': 2}
+_mode_from_name_dict = {'valid': 0,
+                        'same': 1,
+                        'full': 2}
 
 
-def _mode_from_name(mode):
+def _mode_from_name(mode, default_mode):
     if isinstance(mode, str):
-        return _mode_from_name_dict[mode.lower()[0]]
+        if mode.lower() not in _mode_from_name_dict.keys():
+            warnings.warn(
+                "Inexact matches for modes are deprecated, "
+                "please use one of 'valid', 'same', or "
+                "'full' instead.", DeprecationWarning,
+                stacklevel=3)
+            return _mode_from_name_dict[default_mode]
```

Comment:
This is a behavior change - if I pass `'s'`, the expected behavior is:

* Emit the warning
* Pretend I passed `'same'`

What you've implemented here does not do the second bullet point.

Reply:
This makes sense, I'll make the changes 👍🏼

Reply:
What if I pass `'something'`, the expected behaviour should be:
- Emit the warning or an error?

Up till now, `'something'`'s first character, `'s'` would have been checked, which exacts to `'same'`, which is unwanted behaviour. Think we should raise an error instead of warning here?
(For the case in which the first character does not match, it raises an error anyway, refer to the reproducing code at https://github.com/numpy/numpy/issues/17476#issue-716143738)